Scripts/Runtime/Collections/UnsafeBuffer.cs

```diff
@@ -0,0 +1,550 @@
+using System.Runtime.InteropServices;
+using Unity.Assertions;
+using Unity.Burst;
+using Unity.Collections;
+using Unity.Collections.LowLevel.Unsafe;
+using Unity.Jobs;
+using Unity.Jobs.LowLevel.Unsafe;
+
+namespace Anvil.Unity.DOTS.Collections
+{
+    /// <summary>
+    /// Information about a "block" of memory in the buffer.
+    /// Contains the pointer to where the "block" begins and a pointer to another <see cref="BlockInfo"/> instance
+    /// for the next "block" in the linked list. 
+    /// </summary>
+    [BurstCompatible]
+    internal unsafe struct BlockInfo
+    {
+        public static readonly int SIZE = UnsafeUtility.SizeOf<BlockInfo>();
+        public static readonly int ALIGNMENT = UnsafeUtility.AlignOf<BlockInfo>();
+
+        public byte* Data;
+        public BlockInfo* Next;
+    }
+    
+    /// <summary>
+    /// Information about a "lane" which is an indexed sub-section of the buffer that can be read from or written to.
+    /// A lane is commonly used in parallel situations so that each thread stays in its own lane.
+    /// Lane Info preserves the state of writing so that multiple jobs can write to a buffer over time.
+    /// </summary>
+    /// <remarks>
+    /// Unity will refer to this as a "foreachindex" which is a little bit confusing.
+    /// </remarks>
+    [BurstCompatible]
+    [StructLayout(LayoutKind.Sequential, Size = 64)]
+    internal unsafe struct Lane
+    {
+        public BlockInfo* FirstBlock;
+        public BlockInfo* CurrentWriterBlock;
+        public byte* WriterHead;
+        public byte* WriterEndOfBlock;
+        public int Count;
+    }
+    
+    /// <summary>
+    /// Information about the <see cref="UnsafeBuffer{T}"/> itself.
+    /// </summary>
+    [BurstCompatible]
+    internal struct BufferInfo
+    {
+        public int ElementSize;
+        public int ElementAlignment;
+        public int BlockSize;
+        public int LaneCount;
+        public Allocator Allocator;
+    }
+    
+    /// <summary>
+    /// A collection that allows for parallel reading and writing.
+    /// It looks and somewhat behaves similar to a <see cref="UnsafeStream"/> but has some key differences that make it
+    /// more advantageous.
```

Comment:
Hmmmn good question. I can't really think of when I'd want to use `UnsafeStream` vs this right now. There might be determinism aspects that `UnsafeStream` does better but I'm not 100% sure. I've added a TODO about that. 

The one thing that `UnsafeStream` does have an advantage over is if you wanted to write different data to the stream. In that case you obviously wouldn't want to use this since it's constrained to the Type for the elements.

Reply:
I like the edits.